### PR TITLE
Add re2 multi-mode matching

### PR DESCRIPTION
**Pull request by geekeryy**

### CODE REVIEW

**Comment by anuraaga** [comment]
Hi @geekeryy - I'm not sure of the intent of this PR since there isn't a description, but as the goal of this project is to just replace `regexp` package we generally don't add new API. That being said, if there is strong reason to, we can start by adding unstable API to the `experimental` package. To justify it, we need benchmarks comparing using the set API vs creating a concatenated regexp

```go
slices.Sort(exprs)
var grouped []string
for _, e := range exprs {
  grouped = append(grouped, fmt.Sprintf("(%s)", e))
}
re := MustCompile(strings.Join(grouped, "|"))
...
```

AFAIK, internally both are implemented relatively similarly, set is just a light wrapper around `Regexp::Concat` which I think is the same as a `|` operation

https://github.com/google/re2/blob/main/re2/set.cc#L86

So if this transformation is basically equivalent, we don't want to add API for it as users can construct the expression fairly easily.

**Comment by anuraaga** [comment]
Thanks for the example @geekeryy - upon further investigation, I found the performance difference likely comes from forcing DFA mode which can't be done normally otherwise.

https://github.com/google/re2/blob/main/re2/set.cc#L148

So I think this could be a useful feature. Can you rework it to add the new public API to `experimental` instead of the top-level? We may make breaking changes to it.

**Comment by geekeryy** [comment]
> Thanks for the example @geekeryy - upon further investigation, I found the performance difference likely comes from forcing DFA mode which can't be done normally otherwise.
> 
> https://github.com/google/re2/blob/main/re2/set.cc#L148
> 
> So I think this could be a useful feature. Can you rework it to add the new public API to `experimental` instead of the top-level? We may make breaking changes to it.

I have completed it.

**Comment by geekeryy** [comment]
@anuraaga Can you help me merge the code?

**Comment by anuraaga** [comment]
Sorry for the delay. Since this is a big change I will need to spend some time reviewing / helping on it. I'll be able to start tomorrow, it may take some time after that until we get it to merge though.

**Comment by anuraaga** [comment]
@geekeryy Thanks for the patience. I have updated this PR with some changes

- I separated out the cre2 change so the binary can be signed by GitHub https://github.com/wasilibs/go-re2/pull/183
- I simplified the error message handling to remove the buffer. Since cre2 is just an implementation detail for us we can do some specific handling for it
- I renamed the entry point to better match Regexp

Can you help finish this PR by

- Add `String` version of the find method, sharing code in a similar way as we do in Regexp
- Add test cases for various `n`, e.g. negative, `0`, enough for partial match, enough for all match

File: internal/set.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+// Find searches for the first occurrence of any pattern in the Set within the given byte slice.
+// It returns the index of the matched pattern or -1 if no match is found.
+//
+// Parameters:
+// - b: The byte slice to search within.
+//
+// Returns:
+// - int: The index of the matched pattern, or -1 if no match is found.
+func (set *Set) Find(b []byte) int {
```
Comment (anuraaga) [inline internal/set.go:67]: Sorry if my request wasn't clear. Please remove this method and instead add `FindAllString(s string, n int) []int`. It is somewhat similar to

https://github.com/wasilibs/go-re2/blob/main/internal/re2.go#L340

Also note how there we define a private `findAll` method that is shared regardless of if the input is `[]byte` of `string`, this one will be simpler but we should use the similar pattern.

**Review by anuraaga — APPROVED** [review]
Thanks for the patience!

**Pull request merged**